src/rgw/rgw_rest_s3.h

```diff
@@ -663,22 +663,51 @@ static inline int valid_s3_bucket_name(const string& name, bool relaxed=false)
     return -ERR_INVALID_BUCKET_NAME;
   }
 
-  // bucket names must start with a number, letter, or underscore
+  // bucket names must start with a number or letter
   if (!(isalpha(name[0]) || isdigit(name[0]))) {
     if (!relaxed)
       return -ERR_INVALID_BUCKET_NAME;
+      // do we have to support underscore? Maybe needed for backward
+      // compatibility.
     else if (!(name[0] == '_' || name[0] == '.' || name[0] == '-'))
       return -ERR_INVALID_BUCKET_NAME;
   }
 
+  // bucket names cannot end with dash
+  if (name[len-1] == '-')
+    if (!relaxed)
+      return -ERR_INVALID_BUCKET_NAME;
+
   for (const char *s = name.c_str(); *s; ++s) {
     char c = *s;
-    if (isdigit(c) || (c == '.'))
+    if (isdigit(c))
       continue;
-    if (isalpha(c))
-      continue;
-    if ((c == '-') || (c == '_'))
+
+    if (isalpha(c)) {
+      // name cannot contain uppercase letters
+      if (relaxed || islower(c))
+     	continue;
+    }
+
+    if (c == '_')
+      // name cannot contain underscore
+      if (relaxed)
+	continue;
+
+    if (c == '-')
       continue;
+   
+    if (c == '.') {
+      // name cannot have consecutive periods or dashes
+      // adjacent to periods
+      if (!relaxed ) {
+	char p = *(s-1);
```

Comment:
couldn't `s` be the first character here? it doesn't look safe to deference `s-1` without checking

Reply:
Thanks for the review.

As per the amazon spec, seems like bucket names should start with either letter or number. Have added checks at line:666 to bail out for other invalid cases. So ideally we should not reach here if "s" is the first character. But still will add the check to ensure that in case of any change to that assumption in future.